File: server/internal/handlers/identity/workspaces.go
package identityhandler

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/laravel42/berry-circle/server/internal/httpapi"
	"github.com/laravel42/berry-circle/server/internal/identity"
)

type pageInfo struct {
	HasNextPage bool    `json:"hasNextPage"`
	EndCursor   *string `json:"endCursor"`
}

type workspaceResource struct {
	ID          uuid.UUID                  `json:"id"`
	Name        string                     `json:"name"`
	Slug        string                     `json:"slug"`
	Description *string                    `json:"description"`
	Settings    identity.WorkspaceSettings `json:"settings"`
	Role        identity.Role              `json:"role"`
	CreatedAt   string                     `json:"createdAt"`
	UpdatedAt   string                     `json:"updatedAt"`
}

type workspaceConnection struct {
	Nodes    []workspaceResource `json:"nodes"`
	PageInfo pageInfo            `json:"pageInfo"`
}

func (handlers *handlers) listWorkspaces(
	response http.ResponseWriter,
	request *http.Request,
) {
	user := currentUser(request)
	first, after, ok := parseTimePage(
		response,
		request,
		"identity.workspaces."+user.ID.String(),
	)
	if !ok {
		return
	}
	found, err := handlers.service.ListWorkspaces(
		request.Context(),
		user.ID,
		after,
		first+1,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	hasNext := len(found) > first
	if hasNext {
		found = found[:first]
	}
	nodes := make([]workspaceResource, 0, len(found))
	for _, workspace := range found {
		nodes = append(nodes, serializeWorkspace(workspace))
	}
	httpapi.WriteJSON(response, http.StatusOK, workspaceConnection{
		Nodes: nodes,
		PageInfo: pageInfo{
			HasNextPage: hasNext,
			EndCursor:   timeEndCursor("identity.workspaces."+user.ID.String(), found),
		},
	})
}

type createWorkspaceBody struct {
	Name        *string         `json:"name"`
	Slug        *string         `json:"slug"`
	Description json.RawMessage `json:"description"`
}

func (handlers *handlers) createWorkspace(
	response http.ResponseWriter,
	request *http.Request,
) {
	body, raw, ok := decodeBody[createWorkspaceBody](response, request)
	if !ok {
		return
	}
	key, fingerprint, ok := requireIdempotency(response, request, raw)
	if !ok {
		return
	}
	fields := make([]httpapi.FieldError, 0)
	var name, slug string
	if body.Name == nil {
		fields = append(fields, fieldError("/name", "required", "Name is required."))
	} else {
		name = strings.TrimSpace(*body.Name)
		if !validBounded(name, 1, 100) {
			fields = append(fields, fieldError(
				"/name",
				"invalid_length",
				"Name must contain 1 to 100 characters.",
			))
		}
	}
	if body.Slug == nil {
		fields = append(fields, fieldError("/slug", "required", "Slug is required."))
	} else {
		slug = strings.ToLower(strings.TrimSpace(*body.Slug))
		if !validWorkspaceSlug(slug) {
			fields = append(fields, fieldError(
				"/slug",
				"invalid_format",
				"Slug must contain 2 to 50 lowercase letters, digits, or hyphens.",
			))
		}
	}
	description, descriptionSet, descriptionError := parseNullableString(
		body.Description,
		5000,
	)
	if descriptionError != nil {
		fields = append(fields, fieldError(
			"/description",
			"invalid_length",
			"Description must be null or at most 5,000 characters.",
		))
	}
	if len(fields) > 0 {
		writeValidation(response, request, fields...)
		return
	}
	if !descriptionSet {
		description = nil
	}
	workspace, replayed, err := handlers.service.CreateWorkspace(
		request.Context(),
		currentUser(request).ID,
		name,
		slug,
		description,
		key,
		fingerprint,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	if replayed {
		response.Header().Set("Idempotency-Replayed", "true")
	}
	response.Header().Set("Location", "/api/v1/workspaces/"+workspace.ID.String())
	httpapi.WriteJSON(response, http.StatusCreated, serializeWorkspace(workspace))
}

func (handlers *handlers) getWorkspace(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	workspace, err := handlers.service.GetWorkspace(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	httpapi.WriteJSON(response, http.StatusOK, serializeWorkspace(workspace))
}

type workspacePatchBody struct {
	Name        *string         `json:"name"`
	Slug        *string         `json:"slug"`
	Description json.RawMessage `json:"description"`
}

func (handlers *handlers) updateWorkspace(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	body, _, ok := decodeBody[workspacePatchBody](response, request)
	if !ok {
		return
	}
	patch := identity.WorkspacePatch{}
	fields := make([]httpapi.FieldError, 0)
	if body.Name != nil {
		value := strings.TrimSpace(*body.Name)
		if !validBounded(value, 1, 100) {
			fields = append(fields, fieldError(
				"/name",
				"invalid_length",
				"Name must contain 1 to 100 characters.",
			))
		} else {
			patch.Name = &value
		}
	}
	if body.Slug != nil {
		value := strings.ToLower(strings.TrimSpace(*body.Slug))
		if !validWorkspaceSlug(value) {
			fields = append(fields, fieldError(
				"/slug",
				"invalid_format",
				"Slug must contain 2 to 50 lowercase letters, digits, or hyphens.",
			))
		} else {
			patch.Slug = &value
		}
	}
	description, descriptionSet, descriptionError := parseNullableString(
		body.Description,
		5000,
	)
	if descriptionError != nil {
		fields = append(fields, fieldError(
			"/description",
			"invalid_length",
			"Description must be null or at most 5,000 characters.",
		))
	} else if descriptionSet {
		patch.DescriptionSet = true
		patch.Description = description
	}
	if patch.Name == nil && patch.Slug == nil && !patch.DescriptionSet {
		fields = append(fields, fieldError(
			"/",
			"empty_patch",
			"At least one workspace field is required.",
		))
	}
	if len(fields) > 0 {
		writeValidation(response, request, fields...)
		return
	}
	workspace, err := handlers.service.UpdateWorkspace(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
		patch,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	httpapi.WriteJSON(response, http.StatusOK, serializeWorkspace(workspace))
}

func (handlers *handlers) deleteWorkspace(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	if err := handlers.service.DeleteWorkspace(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
	); err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	response.WriteHeader(http.StatusNoContent)
}

func (handlers *handlers) selectWorkspace(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	if !requireEmptyBody(response, request) {
		return
	}
	if err := handlers.service.SelectWorkspace(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
	); err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	workspace, err := handlers.service.GetWorkspace(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	httpapi.WriteJSON(response, http.StatusOK, serializeWorkspace(workspace))
}

func (handlers *handlers) getWorkspaceSettings(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	workspace, err := handlers.service.GetWorkspace(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	httpapi.WriteJSON(response, http.StatusOK, workspace.Settings)
}

type workspaceSettingsPatchBody struct {
	IssuePrefix        *string        `json:"issuePrefix"`
	DefaultRole        *identity.Role `json:"defaultRole"`
	AllowMemberInvites *bool          `json:"allowMemberInvites"`
}

func (handlers *handlers) updateWorkspaceSettings(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	body, _, ok := decodeBody[workspaceSettingsPatchBody](response, request)
	if !ok {
		return
	}
	fields := make([]httpapi.FieldError, 0)
	if body.IssuePrefix != nil {
		value := strings.ToUpper(strings.TrimSpace(*body.IssuePrefix))
		body.IssuePrefix = &value
		if !validIssuePrefix(value) {
			fields = append(fields, fieldError(
				"/issuePrefix",
				"invalid_format",
				"Issue prefix must contain 2 to 12 uppercase letters or digits.",
			))
		}
	}
	if body.DefaultRole != nil &&
		(*body.DefaultRole != identity.RoleMember &&
			*body.DefaultRole != identity.RoleViewer) {
		fields = append(fields, fieldError(
			"/defaultRole",
			"invalid_enum_value",
			"Default role must be member or viewer.",
		))
	}
	if body.IssuePrefix == nil && body.DefaultRole == nil &&
		body.AllowMemberInvites == nil {
		fields = append(fields, fieldError(
			"/",
			"empty_patch",
			"At least one workspace setting is required.",
		))
	}
	if len(fields) > 0 {
		writeValidation(response, request, fields...)
		return
	}
	settings, err := handlers.service.UpdateWorkspaceSettings(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
		identity.WorkspaceSettingsPatch{
			IssuePrefix:        body.IssuePrefix,
			DefaultRole:        body.DefaultRole,
			AllowMemberInvites: body.AllowMemberInvites,
		},
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	httpapi.WriteJSON(response, http.StatusOK, settings)
}

type memberResource struct {
	UserID      uuid.UUID     `json:"userId"`
	WorkspaceID uuid.UUID     `json:"workspaceId"`
	Role        identity.Role `json:"role"`
	Email       string        `json:"email"`
	Name        string        `json:"name"`
	AvatarURL   *string       `json:"avatarUrl"`
	JoinedAt    string        `json:"joinedAt"`
	UpdatedAt   string        `json:"updatedAt"`
}

type memberConnection struct {
	Nodes    []memberResource `json:"nodes"`
	PageInfo pageInfo         `json:"pageInfo"`
}

func (handlers *handlers) listMembers(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	scope := "identity.members." + workspaceID.String()
	first, after, ok := parseNamePage(response, request, scope)
	if !ok {
		return
	}
	found, err := handlers.service.ListMembers(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
		after,
		first+1,
	)
	if err != nil {
		writeDomainError(response, request, err, "Workspace")
		return
	}
	hasNext := len(found) > first
	if hasNext {
		found = found[:first]
	}
	nodes := make([]memberResource, 0, len(found))
	for _, member := range found {
		nodes = append(nodes, serializeMember(member))
	}
	var endCursor *string
	if len(found) > 0 {
		last := found[len(found)-1]
		encoded, err := httpapi.EncodeCursor(scope, identity.NameCursor{
			Name: last.Name,
			ID:   last.UserID,
		})
		if err == nil {
			endCursor = &encoded
		}
	}
	httpapi.WriteJSON(response, http.StatusOK, memberConnection{
		Nodes: nodes,
		PageInfo: pageInfo{
			HasNextPage: hasNext,
			EndCursor:   endCursor,
		},
	})
}

type memberRolePatchBody struct {
	Role *identity.Role `json:"role"`
}

func (handlers *handlers) updateMemberRole(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	targetID, ok := parseID(
		response,
		request,
		chi.URLParam(request, "userId"),
		"Member",
	)
	if !ok {
		return
	}
	body, _, ok := decodeBody[memberRolePatchBody](response, request)
	if !ok {
		return
	}
	if body.Role == nil || !body.Role.Valid() {
		writeValidation(response, request, fieldError(
			"/role",
			"invalid_enum_value",
			"Role must be owner, admin, member, or viewer.",
		))
		return
	}
	member, err := handlers.service.UpdateMemberRole(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
		targetID,
		*body.Role,
	)
	if err != nil {
		writeDomainError(response, request, err, "Member")
		return
	}
	httpapi.WriteJSON(response, http.StatusOK, serializeMember(member))
}

func (handlers *handlers) removeMember(
	response http.ResponseWriter,
	request *http.Request,
) {
	workspaceID, ok := workspaceID(response, request)
	if !ok {
		return
	}
	targetID, ok := parseID(
		response,
		request,
		chi.URLParam(request, "userId"),
		"Member",
	)
	if !ok {
		return
	}
	if err := handlers.service.RemoveMember(
		request.Context(),
		currentUser(request).ID,
		workspaceID,
		targetID,
	); err != nil {
		writeDomainError(response, request, err, "Member")
		return
	}
	response.WriteHeader(http.StatusNoContent)
}

func workspaceID(
	response http.ResponseWriter,
	request *http.Request,
) (uuid.UUID, bool) {
	return parseID(
		response,
		request,
		chi.URLParam(request, "workspaceId"),
		"Workspace",
	)
}

func serializeWorkspace(workspace identity.Workspace) workspaceResource {
	return workspaceResource{
		ID:          workspace.ID,
		Name:        workspace.Name,
		Slug:        workspace.Slug,
		Description: workspace.Description,
		Settings:    workspace.Settings,
		Role:        workspace.Role,
		CreatedAt:   workspace.CreatedAt.UTC().Format(time.RFC3339Nano),
		UpdatedAt:   workspace.UpdatedAt.UTC().Format(time.RFC3339Nano),
	}
}

func serializeMember(member identity.Membership) memberResource {
	return memberResource{
		UserID:      member.UserID,
		WorkspaceID: member.WorkspaceID,
		Role:        member.Role,
		Email:       member.Email,
		Name:        member.Name,
		AvatarURL:   member.AvatarURL,
		JoinedAt:    member.JoinedAt.UTC().Format(time.RFC3339Nano),
		UpdatedAt:   member.UpdatedAt.UTC().Format(time.RFC3339Nano),
	}
}

func timeEndCursor[T interface {
	identity.Workspace | identity.Invitation | identity.PersonalToken
}](scope string, values []T) *string {
	if len(values) == 0 {
		return nil
	}
	var cursor identity.TimeCursor
	switch value := any(values[len(values)-1]).(type) {
	case identity.Workspace:
		cursor = identity.TimeCursor{CreatedAt: value.CreatedAt, ID: value.ID}
	case identity.Invitation:
		cursor = identity.TimeCursor{CreatedAt: value.CreatedAt, ID: value.ID}
	case identity.PersonalToken:
		cursor = identity.TimeCursor{CreatedAt: value.CreatedAt, ID: value.ID}
	}
	encoded, err := httpapi.EncodeCursor(scope, cursor)
	if err != nil {
		return nil
	}
	return &encoded
}

func parseNullableString(
	raw json.RawMessage,
	maximum int,
) (*string, bool, error) {
	if raw == nil {
		return nil, false, nil
	}
	if bytes.Equal(bytes.TrimSpace(raw), []byte("null")) {
		return nil, true, nil
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return nil, true, err
	}
	value = strings.TrimSpace(value)
	if !validBounded(value, 0, maximum) {
		return nil, true, errors.New("string is too long")
	}
	return &value, true, nil
}

func requireEmptyBody(
	response http.ResponseWriter,
	request *http.Request,
) bool {
	body, err := io.ReadAll(http.MaxBytesReader(response, request.Body, 1024))
	if err != nil {
		httpapi.WriteError(
			response,
			request,
			http.StatusRequestEntityTooLarge,
			"PAYLOAD_TOO_LARGE",
			"Request body is too large.",
			nil,
		)
		return false
	}
	if len(bytes.TrimSpace(body)) != 0 {
		writeValidation(response, request, fieldError(
			"/",
			"unrecognized_body",
			"This operation does not accept a request body.",
		))
		return false
	}
	return true
}

File: server-ts/src/mounts/workspaces.ts
import { Hono } from 'hono';
import { requireSession, type AuthVariables } from '../auth/middleware.ts';
import type { SessionService } from '../auth/sessions.ts';
import { json } from '../http/app.ts';
import { ApiError, type FieldError } from '../http/errors.ts';
import { assertValid, decodeBody, fieldError } from '../http/body.ts';
import {
   encodeCursor,
   decodeNameCursor,
   decodeTimeCursor,
   parsePageQuery,
} from '../http/cursor.ts';
import { boundedLength, validIssuePrefix, validWorkspaceSlug } from '../http/validation.ts';
import { domain } from '../identity/errors.ts';
import { validRole, type Role } from '../identity/roles.ts';
import type { Membership, WorkspaceRepository } from '../identity/workspaces.ts';
import type { Workspace, WorkspaceSettings } from '../identity/repository.ts';
import { serializeWorkspace } from './me.ts';
import {
   fingerprintJSON,
   pathId,
   requireEmptyBody,
   requireIdempotencyKey,
   serializeMember,
} from './shared.ts';
import { workspaceInvitationRoutes } from './secrets.ts';
import type { SecretsRepository } from '../identity/secrets.ts';
import type { Mount } from '../http/registry.ts';

/**
 * `/api/v1/workspaces`.
 *
 * Authorization lives in the repository, next to the statements it guards, so
 * a route cannot reach the data without passing it. These handlers validate
 * shape and serialise; they do not decide who may do what.
 */

export interface WorkspaceOptions {
   sessions: SessionService;
   workspaces: WorkspaceRepository;
   /** Invitations live under this prefix, so their routes are added here. */
   secrets: SecretsRepository;
   clock?: () => Date;
}

export function workspaceMounts(options: WorkspaceOptions): Mount[] {
   const route = new Hono<{ Variables: AuthVariables }>();
   route.use('*', requireSession(options.sessions));

   const { workspaces } = options;
   workspaceInvitationRoutes(route, options.secrets, options.clock ?? (() => new Date()));

   route.get('/', async (context) => {
      const userId = context.get('user').id;
      const scope = `identity.workspaces.${userId}`;
      const { first, after } = parsePageQuery(new URL(context.req.url));
      const cursor = after === '' ? null : decodeTimeCursor(after, scope);

      // One more than asked for, so the extra row answers hasNextPage without
      // a second count query.
      const found = await domain('Workspace', () => workspaces.list(userId, cursor, first + 1));
      const { nodes, hasNextPage } = page(found, first);
      return json({
         nodes: nodes.map(serializeWorkspace),
         pageInfo: {
            hasNextPage,
            endCursor: timeCursorOf(scope, nodes),
         },
      });
   });

   route.post('/', async (context) => {
      const userId = context.get('user').id;
      const { value, raw } = await decodeBody<CreateBody>(context, {
         name: 'string',
         slug: 'string',
         description: 'raw',
      });
      const idempotencyKey = requireIdempotencyKey(context.req.raw.headers);

      const fields: FieldError[] = [];
      let name = '';
      let slug = '';
      if (value.name === undefined) {
         fields.push(fieldError('/name', 'required', 'Name is required.'));
      } else {
         name = value.name.trim();
         if (!boundedLength(name, 1, 100)) {
            fields.push(
               fieldError('/name', 'invalid_length', 'Name must contain 1 to 100 characters.')
            );
         }
      }
      if (value.slug === undefined) {
         fields.push(fieldError('/slug', 'required', 'Slug is required.'));
      } else {
         slug = value.slug.trim().toLowerCase();
         if (!validWorkspaceSlug(slug)) {
            fields.push(
               fieldError(
                  '/slug',
                  'invalid_format',
                  'Slug must contain 2 to 50 lowercase letters, digits, or hyphens.'
               )
            );
         }
      }
      const description = nullableString(value.description, 5000, '/description', fields);
      assertValid(fields);

      const { workspace, replayed } = await domain('Workspace', () =>
         workspaces.create({
            actorId: userId,
            name,
            slug,
            description: description.set ? description.value : null,
            idempotencyKey,
            fingerprint: fingerprintJSON(raw),
         })
      );

      const headers: Record<string, string> = {
         Location: `/api/v1/workspaces/${workspace.id}`,
      };
      if (replayed) headers['Idempotency-Replayed'] = 'true';
      const response = json(serializeWorkspace(workspace), 201);
      for (const [name, value] of Object.entries(headers)) response.headers.set(name, value);
      return response;
   });

   route.get('/:workspaceId', async (context) =>
      json(
         serializeWorkspace(
            await domain('Workspace', () =>
               workspaces.get(pathId(context.req.param('workspaceId'), 'Workspace'), context.get('user').id)
            )
         )
      )
   );

   route.patch('/:workspaceId', async (context) => {
      const workspaceId = pathId(context.req.param('workspaceId'), 'Workspace');
      const { value } = await decodeBody<UpdateBody>(context, {
         name: 'string',
         slug: 'string',
         description: 'raw',
      });

      const fields: FieldError[] = [];
      const patch: { name?: string; slug?: string; descriptionSet: boolean; description?: string | null } =
         { descriptionSet: false };
      if (value.name !== undefined) {
         const trimmed = value.name.trim();
         if (!boundedLength(trimmed, 1, 100)) {
            fields.push(
               fieldError('/name', 'invalid_length', 'Name must contain 1 to 100 characters.')
            );
         } else {
            patch.name = trimmed;
         }
      }
      if (value.slug !== undefined) {
         const trimmed = value.slug.trim().toLowerCase();
         if (!validWorkspaceSlug(trimmed)) {
            fields.push(
               fieldError(
                  '/slug',
                  'invalid_format',
                  'Slug must contain 2 to 50 lowercase letters, digits, or hyphens.'
               )
            );
         } else {
            patch.slug = trimmed;
         }
      }
      const description = nullableString(value.description, 5000, '/description', fields);
      if (description.ok && description.set) {
         patch.descriptionSet = true;
         patch.description = description.value;
      }
      // Unlike the profile patch, Go only marks a field present here once it
      // has validated, so an invalid name alone also reads as an empty patch.
      if (patch.name === undefined && patch.slug === undefined && !patch.descriptionSet) {
         fields.push(fieldError('/', 'empty_patch', 'At least one workspace field is required.'));
      }
      assertValid(fields);

      return json(
         serializeWorkspace(
            await domain('Workspace', () =>
               workspaces.update(context.get('user').id, workspaceId, patch)
            )
         )
      );
   });

   route.delete('/:workspaceId', async (context) => {
      await domain('Workspace', () =>
         workspaces.remove(context.get('user').id, pathId(context.req.param('workspaceId'), 'Workspace'))
      );
      return new Response(null, { status: 204 });
   });

   route.post('/:workspaceId/select', async (context) => {
      const workspaceId = pathId(context.req.param('workspaceId'), 'Workspace');
      await requireEmptyBody(context.req.raw);
      const userId = context.get('user').id;
      await domain('Workspace', () => workspaces.select(userId, workspaceId));
      return json(serializeWorkspace(await domain('Workspace', () => workspaces.get(workspaceId, userId))));
   });

   route.get('/:workspaceId/settings', async (context) => {
      const workspace = await domain('Workspace', () =>
         workspaces.get(pathId(context.req.param('workspaceId'), 'Workspace'), context.get('user').id)
      );
      return json(serializeSettings(workspace.settings));
   });

   route.patch('/:workspaceId/settings', async (context) => {
      const workspaceId = pathId(context.req.param('workspaceId'), 'Workspace');
      const { value } = await decodeBody<SettingsBody>(context, {
         issuePrefix: 'string',
         defaultRole: 'string',
         allowMemberInvites: 'boolean',
      });

      const fields: FieldError[] = [];
      const patch: { issuePrefix?: string; defaultRole?: string; allowMemberInvites?: boolean } = {};
      if (value.issuePrefix !== undefined) {
         const normalised = value.issuePrefix.trim().toUpperCase();
         patch.issuePrefix = normalised;
         if (!validIssuePrefix(normalised)) {
            fields.push(
               fieldError(
                  '/issuePrefix',
                  'invalid_format',
                  'Issue prefix must contain 2 to 12 uppercase letters or digits.'
               )
            );
         }
      }
      if (value.defaultRole !== undefined) {
         patch.defaultRole = value.defaultRole;
         // Only member and viewer: a workspace that admitted everyone as an
         // admin by default would have no boundary left to enforce.
         if (value.defaultRole !== 'member' && value.defaultRole !== 'viewer') {
            fields.push(
               fieldError('/defaultRole', 'invalid_enum_value', 'Default role must be member or viewer.')
            );
         }
      }
      if (value.allowMemberInvites !== undefined) patch.allowMemberInvites = value.allowMemberInvites;
      if (
         value.issuePrefix === undefined &&
         value.defaultRole === undefined &&
         value.allowMemberInvites === undefined
      ) {
         fields.push(fieldError('/', 'empty_patch', 'At least one workspace setting is required.'));
      }
      assertValid(fields);

      return json(
         serializeSettings(
            await domain('Workspace', () =>
               workspaces.updateSettings(context.get('user').id, workspaceId, patch)
            )
         )
      );
   });

   route.get('/:workspaceId/members', async (context) => {
      const workspaceId = pathId(context.req.param('workspaceId'), 'Workspace');
      const scope = `identity.members.${workspaceId}`;
      const { first, after } = parsePageQuery(new URL(context.req.url));
      const cursor = after === '' ? null : decodeNameCursor(after, scope);

      const found = await domain('Workspace', () =>
         workspaces.listMembers(context.get('user').id, workspaceId, cursor, first + 1)
      );
      const { nodes, hasNextPage } = page(found, first);
      const last = nodes.at(-1);
      return json({
         nodes: nodes.map(serializeMember),
         pageInfo: {
            hasNextPage,
            endCursor: last ? encodeCursor(scope, { name: last.name, id: last.userId }) : null,
         },
      });
   });

   route.patch('/:workspaceId/members/:userId', async (context) => {
      const workspaceId = pathId(context.req.param('workspaceId'), 'Workspace');
      const targetId = pathId(context.req.param('userId'), 'Member');
      const { value } = await decodeBody<{ role?: string }>(context, { role: 'string' });
      if (value.role === undefined || !validRole(value.role)) {
         assertValid([
            fieldError('/role', 'invalid_enum_value', 'Role must be owner, admin, member, or viewer.'),
         ]);
      }

      return json(
         serializeMember(
            await domain('Member', () =>
               workspaces.updateMemberRole(
                  workspaceId,
                  context.get('user').id,
                  targetId,
                  value.role as Role
               )
            )
         )
      );
   });

   route.delete('/:workspaceId/members/:userId', async (context) => {
      await domain('Member', () =>
         workspaces.removeMember(
            pathId(context.req.param('workspaceId'), 'Workspace'),
            context.get('user').id,
            pathId(context.req.param('userId'), 'Member')
         )
      );
      return new Response(null, { status: 204 });
   });

   return [{ prefix: '/api/v1/workspaces', handler: route }];
}

interface CreateBody {
   name?: string;
   slug?: string;
   description?: unknown;
}
type UpdateBody = CreateBody;

interface SettingsBody {
   issuePrefix?: string;
   defaultRole?: string;
   allowMemberInvites?: boolean;
}

/** Trims the over-fetched row and reports whether it existed. */
function page<T>(found: T[], first: number): { nodes: T[]; hasNextPage: boolean } {
   const hasNextPage = found.length > first;
   return { nodes: hasNextPage ? found.slice(0, first) : found, hasNextPage };
}

function timeCursorOf(scope: string, nodes: Workspace[]): string | null {
   const last = nodes.at(-1);
   return last ? encodeCursor(scope, { createdAt: last.createdAt, id: last.id }) : null;
}


/**
 * `null` clears, an absent field leaves it, and anything else must be a
 * bounded string. Ported from parseNullableString.
 */
function nullableString(
   raw: unknown,
   maximum: number,
   path: string,
   fields: FieldError[]
): { ok: boolean; set: boolean; value: string | null } {
   if (raw === undefined) return { ok: true, set: false, value: null };
   if (raw === null) return { ok: true, set: true, value: null };
   if (typeof raw !== 'string' || !boundedLength(raw.trim(), 0, maximum)) {
      fields.push(
         fieldError(path, 'invalid_length', `Description must be null or at most 5,000 characters.`)
      );
      return { ok: false, set: true, value: null };
   }
   return { ok: true, set: true, value: raw.trim() };
}

function serializeSettings(settings: WorkspaceSettings): Record<string, unknown> {
   return {
      issuePrefix: settings.issuePrefix,
      defaultRole: settings.defaultRole,
      allowMemberInvites: settings.allowMemberInvites,
   };
}

